system-configuration/src/network_configuration.rs

```diff
@@ -0,0 +1,468 @@
+    pub fn name(&self) -> String {
+        unsafe { CFString::wrap_under_get_rule(SCNetworkServiceGetName(self.0)) }.to_string()
+    }
+
+    /// Returns this network service is enabled or disabled.
+    pub fn enabled(&self) -> bool {
+        let ret = unsafe { SCNetworkServiceGetEnabled(self.0) };
+        ret == 1
+    }
+
+    /// Returns the DNS infomation on this network service
+    pub fn dns(&self) -> SCNetworkServiceDns {
+        let store = SCDynamicStoreBuilder::new("ns_dns").build();
+
+        let query = |path: String| -> (Option<String>, Option<Vec<IpAddr>>) {
+            let mut _domain_name: Option<String> = None;
+            let mut _server_addresses: Option<Vec<IpAddr>> = None;
+
+            if let Some(value) = store.get(CFString::new(&path)) {
+                if let Some(dict) = value.downcast_into::<CFDictionary>() {
+                    if let Some(domain_name) =
+                        dict.find2(&CFString::from_static_string("DomainName"))
+                    {
+                        let domain_name = unsafe { CFType::wrap_under_get_rule(domain_name) };
+                        if let Some(domain_name) = domain_name.downcast::<CFString>() {
+                            _domain_name = Some(domain_name.to_string());
+                        }
+                    }
+
+                    if let Some(addrs) =
+                        dict.find2(&CFString::from_static_string("ServerAddresses"))
+                    {
+                        let addrs = unsafe { CFType::wrap_under_get_rule(addrs) };
+                        if let Some(addrs) = addrs.downcast::<CFArray<CFString>>() {
+                            let mut temp = Vec::new();
+                            for addr in addrs.iter() {
+                                if let Ok(ip_addr) = addr.to_string().parse::<IpAddr>() {
+                                    temp.push(ip_addr);
+                                }
+                            }
+
+                            if temp.len() > 0 {
+                                _server_addresses = Some(temp);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return (_domain_name, _server_addresses);
+        };
+
+        let (state_domain_name, state_server_addresses) =
+            query(format!("State:/Network/Service/{}/DNS", self.id()));
+        let (setup_domain_name, setup_server_addresses) =
+            query(format!("Setup:/Network/Service/{}/DNS", self.id()));
+
+        SCNetworkServiceDns {
+            state_domain_name: state_domain_name,
+            state_server_addresses: state_server_addresses,
+            setup_domain_name: setup_domain_name,
+            setup_server_addresses: setup_server_addresses,
+        }
+    }
+
+    /// Setting DNS on this network service
+    pub fn set_dns(&self, dns: SCNetworkServiceDns) -> bool {
```

Comment:
This method only sets the `Setup` part of the DNS, never the `State`. Since a `SCNetworkServiceDns` is passed in one could easily think it would set both. And there is also the possible confusion about if a `setup_server_addresses: None` should not change the addresses at all (current implementation) or change it into no addresses.

Expressing "update structs", structs that can conditionally update some fields, can be hard. And it usually ends up with multiple layers of `Option<...>`. To avoid that and take a more direct approach I'm wondering if the following might be a desirable way?:
```rust
set_dns_setup_domain_name(&mut self, domain_name: Option<String>)
set_dns_state_domain_name(&mut self, domain_name: Option<String>)

set_dns_setup_addresses(&mut self, addresses: Option<Vec<IpAddr>>)
set_dns_state_addresses(&mut self, addresses: Option<Vec<IpAddr>>)
```
Where a `None` argument in both cases *removes* the corresponding "ServerAddresses"/"DomainName" key from the store.

What do you think @pronebird?